### PR TITLE
Fix infinite loop

### LINKED ISSUE #512: Infinite loop while input object pointing at each other
https://github.com/Artem-Schander/graphql-infinite-loop

### PR DESCRIPTION
**Pull request by crissi**
## Summary
Fixes #512

Fix the infinite loop as well as sending the correct matching input data to the rule-callback.

If people have been using rule-callback in a nested input object, the input to the object was before always the full input, now the input data matches the level of nesting.

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### CODE REVIEW

**Comment by mfn** [comment]
Ok! That one looks massive…
will take me some time to dig through it; thanks!

**Comment by mfn** [comment]
Sorry for the long wait, I've looked into the high level functionality: very cool, totally want to merge this!

Going now through the code for giving detailed feedback.

**Comment by crissi** [comment]
All cool, take your time😊

Hent Outlook til iOS<https://aka.ms/o0ukef>
________________________________
Fra: Markus Podar <notifications@github.com>
Sendt: Saturday, January 25, 2020 8:46:00 PM
Til: rebing/graphql-laravel <graphql-laravel@noreply.github.com>
Cc: Christian <c2n@live.dk>; Author <author@noreply.github.com>
Emne: Re: [rebing/graphql-laravel] Fix infinite loop (#579)


Sorry for the long wait, I've looked into the high level functionality: very cool, totally want to merge this!

Going now through the code for giving detailed feedback.

File: tests/Support/Objects/UpdateExampleMutationWithInputType.php [hunk 1]
```diff
@@ -49,7 +49,6 @@ public function args(): array
             ],
 
             'test_with_rules_nullable_input_object' => [
-                'name' => 'test',
```
Comment (mfn) [inline tests/Support/Objects/UpdateExampleMutationWithInputType.php]: Why?

File: tests/Support/Objects/ExampleValidationInputObject.php [hunk 2]
```diff
@@ -24,10 +24,15 @@ public function fields(): array
     {
         return [
             'val' => [
-                'name' => 'val',
```
Comment (mfn) [inline tests/Support/Objects/ExampleValidationInputObject.php]: Why?

File: tests/Support/Objects/UpdateExampleMutationWithInputType2.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+declare(strict_types=1);
+
+namespace Rebing\GraphQL\Tests\Support\Objects;
+
+use GraphQL\Type\Definition\Type;
+use PHPUnit\Framework\Assert;
+use Rebing\GraphQL\Support\Facades\GraphQL;
+use Rebing\GraphQL\Support\Mutation;
+
+class UpdateExampleMutationWithInputType2 extends Mutation
```
Comment (mfn) [inline tests/Support/Objects/UpdateExampleMutationWithInputType2.php]: `2`, same feedback as the other one

File: Readme.md [hunk 4]
```diff
@@ -446,7 +446,13 @@ class UpdateUserEmailMutation extends Mutation
     {
         return [
             'id' => ['required'],
-            'email' => ['required', 'email']
+            'email' => ['required', 'email'],
+            'password' => function($arguments, $allArguments) {
```
Comment (mfn) [inline Readme.md]: `$arguments, $allArguments`

I guess this comes from `\Rebing\GraphQL\Support\Rules::resolveRules`

Do we have better names for this (also through the code / tests), this is quite ambiguous to say at least.

File: tests/Unit/MutationTest.php [hunk 5]
```diff
@@ -123,6 +92,36 @@ public function testValidationError(): void
 
         $attributes = $field->getAttributes();
 
+        try {
+            $attributes['resolve'](null, [
+                'test_with_rules_non_nullable_input_object' => [
+                    'val' => 4,
+                ],
+            ], [], null);
+        } catch (ValidationError $e) {
```
Comment (mfn) [inline tests/Unit/MutationTest.php]: This pattern is prone to be miss if the underlying code changes/breaks.

I recommend this:
```php
/** @var ValidationError|null $exc */
$exc = null;
try {
  …
} catch (ValidationError $exc) {
  // Deliberately do nothing
}
$this->assertInstanceOf(ValidationError::class, $exc);
// your regular assertions
```

File: tests/Unit/MutationTest.php [hunk 6]
```diff
@@ -123,6 +92,36 @@ public function testValidationError(): void
...
+    public function testWithEmptyInput(): void
+    {
+        $class = $this->getFieldClass();
+        $field = new $class();
+
+        $attributes = $field->getAttributes();
+
         try {
             $attributes['resolve'](null, [], [], null);
         } catch (ValidationError $e) {
```
Comment (mfn) [inline tests/Unit/MutationTest.php]: I see. The feedback from the last test: this pattern was already here.

Please fix it here, too.

File: tests/Unit/MutationTest.php [hunk 7]
```diff
@@ -131,15 +130,133 @@ public function testValidationError(): void
...
+        $class = $this->getFieldClass();
+        $field = new $class();
+
+        $attributes = $field->getAttributes();
+
+        try {
+            $attributes['resolve'](null, [
+                'test_with_rules' => 'test',
+            ], [], null);
+        } catch (ValidationError $e) {
```
Comment (mfn) [inline tests/Unit/MutationTest.php]: Please change the pattern here

File: tests/Unit/MutationTest.php [hunk 8]
```diff
@@ -131,15 +130,133 @@ public function testValidationError(): void
...
+        $field = new $class();
+
+        $attributes = $field->getAttributes();
+
+        try {
+            $attributes['resolve'](null, [
+                'test_with_rules_non_nullable_input_object' => [],
+                'test_with_rules_nullable_input_object' => [],
+            ], [], null);
+        } catch (ValidationError $e) {
```
Comment (mfn) [inline tests/Unit/MutationTest.php]: Please change the pattern here

File: tests/Unit/MutationTest.php [hunk 9]
```diff
@@ -131,15 +130,133 @@ public function testValidationError(): void
...
+        $class = $this->getFieldClass();
+        $field = new $class();
+
+        $attributes = $field->getAttributes();
+
+        try {
+            $attributes['resolve'](null, [
+                'test_with_rules_non_nullable_list_of_non_nullable_input_object' => [],
+            ], [], null);
+        } catch (ValidationError $e) {
```
Comment (mfn) [inline tests/Unit/MutationTest.php]: Please change the pattern here

File: tests/Unit/MutationTest.php [hunk 10]
```diff
@@ -131,15 +130,133 @@ public function testValidationError(): void
...
+
+        try {
+            $attributes['resolve'](null, [
+                'test_with_rules_non_nullable_list_of_non_nullable_input_object' => [
+                    [
+                        'val' => 1245,
+                    ],
+                ],
+            ], [], null);
+        } catch (ValidationError $e) {
```
Comment (mfn) [inline tests/Unit/MutationTest.php]: Please change the pattern here

File: tests/Unit/MutationTest.php [hunk 11]
```diff
@@ -169,4 +286,19 @@ public function testCustomValidationErrorMessages(): void
...
+        $attributes = $field->getAttributes();
+
+        try {
+            $attributes['resolve'](null, [
+                'test_with_rules_callback_params' => [
+                    'otherValue' => 1337,
+                ],
+            ], [], null);
+        } catch (ValidationError $e) {
+        }
```
Comment (mfn) [inline tests/Unit/MutationTest.php]: If we don't care about the exceptions details (beyond the message), I recommend using the built-in phpunit-way: `$this->expectException(…);`

File: tests/Support/Objects/UpdateExampleMutationWithInputType.php (hunk 1 above)
Comment (crissi) [reply tests/Support/Objects/UpdateExampleMutationWithInputType.php]: don't know why they are there in the first place? does the name attribute do anything?

Comment (mfn) [reply tests/Support/Objects/UpdateExampleMutationWithInputType.php]: > does the name attribute to anything

Well, yes 😄

You can use it to change the name from the key:
- key is a string, use an actual different name
- no explicit key (i.e. plain array), in which case the `name` is actually required

Back to this code: sure, might not be in use/make sense here but my point is: a PR should be focused and random clean ups should not be part of bigger feature like this.

We can totally remove all the `name` which aren't in use, probably a copypasta galore, but then it should be done a) consequently and b) separate from this PR.

**Comment by crissi** [comment]
updated

**Comment by mfn** [comment]
https://github.com/rebing/graphql-laravel/releases/tag/5.0.0 has been released which includes this!

**Pull request merged**